src/topology/locally_constant/basic.lean

```diff
@@ -0,0 +1,260 @@
+
+This file setups the theory of locally constant function from a topological space to a type.
+
+## Main definitions and constructions
+
+* `is_locally_constant f` : a map `f : X → Y` where `X` is a topological space is locally
+                            constant if everything set in `Y` has an open preimage.
+* `locally_constant X Y` : the type of locally constant maps from `X` to `Y`
+* `locally_constant.map` : push-forward of locally constant maps
+* `locally_constant.comap` : pull-back of locally constant maps
+
+-/
+
+variables {X Y Z α : Type*} [topological_space X]
+
+open_locale topological_space
+
+/-- A function between topological spaces is locally constant if the preimage of any set is open. -/
+def is_locally_constant (f : X → Y) : Prop := ∀ s : set Y, is_open (f ⁻¹' s)
+
+namespace is_locally_constant
+
+lemma exists_open {f : X → Y} (hf : is_locally_constant f) (x : X) :
+  ∃ (U : set X) (hU : is_open U) (hx : x ∈ U), ∀ x' ∈ U, f x' = f x :=
+⟨f ⁻¹' {(f x)}, hf _, set.mem_singleton _, λ x' hx', set.mem_singleton_iff.mp hx'⟩
+
+lemma exists_nhds {f : X → Y} (hf : is_locally_constant f) (x : X) :
+  ∃ U ∈ 𝓝 x, ∀ x' ∈ U, f x' = f x :=
+let ⟨U, hU, hx, H⟩ := hf.exists_open x in ⟨U, mem_nhds_sets hU hx, H⟩
+
+lemma iff_exists_open (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ (U : set X) (hU : is_open U) (hx : x ∈ U), ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_open, _⟩,
+  assume h s,
+  rw is_open_iff_forall_mem_open,
+  assume x hx,
+  obtain ⟨U, hU, hxU, H⟩ := h x,
+  refine ⟨U, _, hU, hxU⟩,
+  assume x' hx',
+  simp only [*, set.mem_preimage] at *,
+end
+
+lemma iff_exists_nhds (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ U ∈ 𝓝 x, ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_nhds, _⟩,
+  assume h,
+  rw iff_exists_open,
+  assume x,
+  obtain ⟨U, hU, H⟩ := h x,
+  obtain ⟨V, hVU, hV, hxV⟩ : ∃ (V : set X) (H : V ⊆ U), is_open V ∧ x ∈ V,
+  by rwa mem_nhds_sets_iff at hU,
+  refine ⟨V, hV, hxV, _⟩,
+  assume x' hx',
+  solve_by_elim only [H, hxV, hx', hVU]
+end
+
+lemma of_constant (f : X → Y) (h : ∃ y, ∀ x, f x = y) :
+  is_locally_constant f :=
+begin
+  obtain ⟨y, hy⟩ := h,
+  rw iff_exists_nhds,
+  intro x,
+  refine ⟨set.univ, filter.univ_mem_sets, _⟩,
+  rintro x -,
+  rw [hy, hy]
+end
+
+lemma const (y : Y) : is_locally_constant (function.const X y) :=
+of_constant _ ⟨y, λ _, rfl⟩
+
+protected lemma continuous {_ : topological_space Y} {f : X → Y} (hf : is_locally_constant f) :
+  continuous f :=
+⟨λ U hU, hf _⟩
+
+lemma iff_continuous {_ : topological_space Y} [discrete_topology Y] (f : X → Y) :
+  is_locally_constant f ↔ continuous f :=
+⟨is_locally_constant.continuous, λ h s, h.is_open_preimage s (is_open_discrete _)⟩
+
+lemma iff_continuous_bot (f : X → Y) :
+  is_locally_constant f ↔ @continuous X Y _ ⊥ f :=
+iff_continuous f
+
+lemma apply_eq_of_is_preconnected {f : X → Y} (hf : is_locally_constant f)
+  (s : set X) (hs : is_preconnected s) (x y : X) (hx : x ∈ s) (hy : y ∈ s) :
+  f y = f x :=
+begin
```

Comment:
You don't need `is_preconnected.image` or a topology on `Y` here. It's easier to apply the definition of `is_preconnected` to `tx = {z | f z = f x}` and `ty = {z | f z = f y}`.

Reply:
Fixed.